csharp/src/Google.Protobuf/CodedInputReader.cs

```diff
@@ -0,0 +1,879 @@
+                    if (value < 128)
+                    {
+                        result |= value << 14;
+                        tag = (uint)result;
+                    }
+                    else
+                    {
+                        result = (value & 0x7f) << 14;
+                        ThrowEndOfInputIfFalse(reader.TryRead(out value));
+                        result |= value << 21;
+                        tag = (uint)result;
+                    }
+                }
+            }
+
+            lastTag = tag;
+
+            if (WireFormat.GetTagFieldNumber(tag) == 0)
+            {
+                // If we actually read a tag with a field of 0, that's not a valid tag.
+                throw InvalidProtocolBufferException.InvalidTag();
+            }
+            if (ReachedLimit)
+            {
+                return 0;
+            }
+            return tag;
+        }
+
+        /// <summary>
+        /// Skips the data for the field with the tag we've just read.
+        /// This should be called directly after <see cref="ReadTag"/>, when
+        /// the caller wishes to skip an unknown field.
+        /// </summary>
+        /// <remarks>
+        /// This method throws <see cref="InvalidProtocolBufferException"/> if the last-read tag was an end-group tag.
+        /// If a caller wishes to skip a group, they should skip the whole group, by calling this method after reading the
+        /// start-group tag. This behavior allows callers to call this method on any field they don't understand, correctly
+        /// resulting in an error if an end-group tag has not been paired with an earlier start-group tag.
+        /// </remarks>
+        /// <exception cref="InvalidProtocolBufferException">The last tag was an end-group tag</exception>
+        /// <exception cref="InvalidOperationException">The last read operation read to the end of the logical input</exception>
+        public void SkipLastField()
+        {
+            if (lastTag == 0)
+            {
+                throw new InvalidOperationException("SkipLastField cannot be called at the end of a input");
+            }
+            switch (WireFormat.GetTagWireType(lastTag))
+            {
+                case WireFormat.WireType.StartGroup:
+                    SkipGroup(lastTag);
+                    break;
+                case WireFormat.WireType.EndGroup:
+                    throw new InvalidProtocolBufferException(
+                        "SkipLastField called on an end-group tag, indicating that the corresponding start-group was missing");
+                case WireFormat.WireType.Fixed32:
+                    ReadFixed32();
+                    break;
+                case WireFormat.WireType.Fixed64:
+                    ReadFixed64();
+                    break;
+                case WireFormat.WireType.LengthDelimited:
+                    var length = ReadLength();
+                    ThrowEndOfInputIfFalse(reader.Remaining >= length);
```

Comment:
I think you need to consult  the currentLimit too? Looks like without it you can go over the allowed limit for current message without noticing?
I suggest reintroducing SkipRawBytes(int) method to avoid missing this.